prow/cache/cache.go

```diff
@@ -0,0 +1,220 @@
+		//  2. actually construct the value (expensive operation);
+		//  3. resolve the promise to alert all threads looking at the same Promise
+		//     get the value from Step 2.
+		//
+		// This mitigation strategy is a kind of "duplicate suppression", also
+		// called "request coalescing". The problem of having to deal with a
+		// flood of multiple requests for the same cache entry is also called
+		// "cache stampede".
+
+		// Step 1
+		//
+		// Let other threads know about our promise to construct the value. We
+		// don't care if the underlying LRU cache had to evict an existing
+		// entry.
+		promise = newPromise(valConstructor)
+		_ = lruCache.Add(key, promise)
+		// We must unlock here so that the cache does not block other GetOrAdd()
+		// calls to it for different (or same) key/value pairs.
+		lruCache.Unlock()
+
+		// Step 2 & 3
+		//
+		// Construct the value (expensive operation), and broadcast to all
+		// watchers of this promise that it is ready to be read from (no data
+		// race!).
+		promise.resolve()
+
+		// If the value construction (expensive operation) failed, then we
+		// delete the cached entry so that we may attempt to re-try again in the
+		// future (instead of waiting for the LRUCache to evict it on its own
+		// over time).
+		//
+		// NOTE: It may be the case that the underlying lruCache itself decided
+		// to evict this key by the time we try to Lock() it here and evict it
+		// ourselves. I.e., it may be the case that the lruCache evicted our key
+		// because there just happened to be a massive load of calls with lots
+		// of different keys, forcing all old cached values to be evicted. But
+		// this is a minor concern because (1) it is unlikely to happen and (2)
+		// even if it does happen, our eviction will be a NOP because the key we
+		// want to delete wouldn't be in the cache anyway (it's already been
+		// evicted!).
+		//
+		// Another possibility is that by the time we run attempt to delete the
+		// key here, there has been not only an eviction of this same key, but
+		// the creation of another entry with the same key with valid results.
+		// So at worst we would be wrongfully invalidating a cache entry.
+		//
+		// TODO: If our cache implementation supports a TTL mechanism, then we
+		// could just set that instead and let the cached entry expire on its
+		// own (we would not have to do this eviction ourselves manually).
+		if promise.err != nil {
+			logrus.WithField("key", key).Infof("promise was successfully resolved, but the call to resolve() returned an error; deleting key from cache...")
+			lruCache.Lock()
+			keyWasFoundBeforeRemoval := lruCache.Remove(key)
+			lruCache.Unlock()
+			if keyWasFoundBeforeRemoval {
+				logrus.WithField("key", key).Infof("successfully deleted")
+			} else {
+				err := fmt.Errorf("unexpected race: key deleted by the cache without our knowledge; our own deletion of this key was a NOP")
+				logrus.WithField("key", key).Error(err)
```

Comment:
This could happen during normal operation (although very unlikely) and doesn't constitute a problem. Please lower this to at least info level and indicate that this is not problematic in the message.

Reply:
done